### PR TITLE
Correctly specify dependency on typing_extensions

### PR DESCRIPTION
**Pull request by AustinT**
Syntheseus previously depended on typing_extensions without specifying this in the `pyproject.toml` file, causing a fresh install to fail the tests. This PR fixes this by:

1. Wrapping the second import of typing_extensions in an if/else block to only run if python < 3.8 is used (otherwise an import from `typing` can be used directly). This was already present in the first import of typing_extensions.
2. Modify pyproject.toml to specify that `typing_extensions` is required, but _only if python < 3.8_.

### CODE REVIEW

**Comment by AustinT** [comment]
(I noticed this issue when investigating #106 )

**Comment by AustinT** [comment]
@kmaziarz the tests are failing in this PR due to a segmentation fault, but this doesn't seem very related to the PR? I've tried re-running 3 times now but it still won't pass. Not sure what's going on here...

**Comment by kmaziarz** [comment]
> Modify pyproject.toml to specify that `typing_extensions` is required, but only if python < 3.8

My thinking was that we should drop support for Python 3.7 soon anyway; we only maintained it due to being required by GLN, but as of #103 this is no longer the case.

Perhaps we should release `0.5.0` to have a final numbered version that still supports Python 3.7, and then bump the version requirement after that?

> Tests are failing in this PR due to a segmentation fault, but this doesn't seem very related to the PR

Yeah, doesn't seem related to the PR. It also happens from `main`, and seems to also be caused by `torch-scatter`... 😆 I'll need to take a closer look at what's going on

**Comment by kmaziarz** [comment]
Update: I don't think the environment issues are _really_ caused by `torch_scatter`, and rather by the fact that installing `syntheseus[all]` now reinstalls the `torch` despite it already being present in the env. This is due to another package updating upstream. I'll send a fix soon although we may want to set things up a bit more robustly in the long term...

**Comment by AustinT** [comment]
Great Kris, thanks for looking into this! I agree about not supporting python 3.7 anymore- that makes our lives a lot easier. Releasing a final python 3.7 version and then dropping support sounds good.

**Comment by kmaziarz** [comment]
The issue should be fixed as of #108. I think this is not everything, as I also get problems with building `torch-scatter` as reported in the other issue. I'll look into that too, but I think that's separate, and doesn't seem to happen in CI. At the very least #108 should unblock your PR, which I imagine we still may want to merge in order to get a fully working `0.5.0` version (before we drop support for Python 3.7 and strip out some of the changes from your PR)?

**Comment by AustinT** [comment]
Sounds good, once #108 is merged lets re-run CI, merge this if there are no issues, make a new release, then drop python 3.7 support!

**Comment by kmaziarz** [comment]
> Sounds good, once #108 is merged lets re-run CI, merge this if there are no issues, make a new release, then drop python 3.7 support!

I've rebased the PR onto `main` now; lets see if the CI passes this time around.

**Comment by AustinT** [comment]
Looks like it passes @kmaziarz. Should I merge?

**Review by kmaziarz — APPROVED** [review]

**Comment by kmaziarz** [comment]
Seems good to merge, maybe just add a CHANGELOG entry.

**Comment by AustinT** [comment]
Sorry, didn't see your comment re CHANGELOG

**Comment by AustinT** [comment]
@kmaziarz fixed CHANGELOG in #109

**Pull request merged**
